chains/evm/contracts/onRamp/OnRamp.sol

```diff
@@ -568,25 +574,42 @@ contract OnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, Ownable2StepMsgSender
       revert UnsupportedToken(tokenAndAmount.token);
     }
 
-    // TODO support CCIP_POOL_V2
+    Pool.LockOrBurnOutV1 memory poolReturnData;
+    uint256 destTokenAmount;
 
-    Pool.LockOrBurnOutV1 memory poolReturnData = sourcePool.lockOrBurn(
-      Pool.LockOrBurnInV1({
-        receiver: receiver,
-        remoteChainSelector: destChainSelector,
-        originalSender: originalSender,
-        amount: tokenAndAmount.amount,
-        localToken: tokenAndAmount.token
-      })
-    );
+    Pool.LockOrBurnInV1 memory lockOrBurnInput = Pool.LockOrBurnInV1({
+      receiver: receiver,
+      remoteChainSelector: destChainSelector,
+      originalSender: originalSender,
+      amount: tokenAndAmount.amount,
+      localToken: tokenAndAmount.token
+    });
+
+    // If the pool declares support for IPoolV2, it can handle `finality` and `tokenArgs`.
+    // Use the V2 overload which returns a potentially adjusted destination amount.
+    if (IERC165(address(sourcePool)).supportsInterface(type(IPoolV2).interfaceId)) {
+      (poolReturnData, destTokenAmount) =
+        IPoolV2(address(sourcePool)).lockOrBurn(lockOrBurnInput, blockConfirmationRequested, tokenArgs);
+    } else {
+      // V1 pools don't understand `blockConfirmationRequested`/`tokenArgs`.
+      // We enforce default for `blockConfirmationRequested` and no `tokenArgs` to avoid silent mis-interpretation.
+      if (blockConfirmationRequested != 0) {
+        revert CustomBlockConfirmationNotSupportedOnPoolV1();
+      }
+      if (tokenArgs.length != 0) {
+        revert TokenArgsNotSupportedOnPoolV1();
+      }
+      poolReturnData = sourcePool.lockOrBurn(lockOrBurnInput);
+      // V1 returns only LockOrBurnOutV1, the destination amount is 1:1 with the source amount.
+      destTokenAmount = tokenAndAmount.amount;
```

Comment:
maybe nicer to move this to the declaration and just have V2 override it if needed. Slightly cleaner code imo